File: test-projects/ScalaTest/src/test/scala/SeedNetworkTest.scala
import minitest._
import synapses.Library._

object SeedNetworkTest extends SimpleTestSuite {

  val layers = List(4, 6, 5, 3)

  val neuralNetwork: NeuralNetwork =
    NeuralNetwork.initWithSeed(1000, layers)

  val neuralNetworkJson = """[[{"activationF":"sigmoid","weights":[-0.8928384189584146,0.9211880913772268,0.027180243817137795,0.10852652658514339,-0.20162813099768573]},{"activationF":"sigmoid","weights":[-0.3161167802991376,-0.9489930079469029,-0.26015677306584273,-0.6978873003833059,0.2874994065396639]},{"activationF":"sigmoid","weights":[-0.21496286920606544,-0.9357227175449083,-0.9155202030305005,-0.9129309852279106,-0.6978999469719491]},{"activationF":"sigmoid","weights":[0.25806133542967125,0.47588465535587576,-0.5681549645809776,0.9815364553014789,-0.6175472916357287]},{"activationF":"sigmoid","weights":[0.8131999422319118,-0.18065888278217646,-0.8943132277876955,0.294939737903549,-0.13946653126709108]},{"activationF":"sigmoid","weights":[0.3867614051417534,0.4437184379986423,0.2690674140331535,0.32810615530021336,0.17323902449022643]}],[{"activationF":"sigmoid","weights":[-0.1496727007713341,-0.10075233916843396,0.7276149322065253,0.8063910194692945,0.6688055519864093,-0.975106575568284,0.09221847729601618]},{"activationF":"sigmoid","weights":[-0.9446600359811137,-0.5964794872043058,0.7903102784527913,-0.9051888424151342,-0.09122610135386999,0.6535569969369026,-0.22748365874317633]},{"activationF":"sigmoid","weights":[0.09017935830009893,0.6208093403826789,0.33538576542866516,0.753380386023317,-0.2055709809859254,0.9488128412675767,-0.07338684978278609]},{"activationF":"sigmoid","weights":[-0.07576012821021783,-0.35625015691520523,-0.06294831714455151,-0.5599640923034961,-0.35867440290460384,0.5199823465293354,0.6691922308539804]},{"activationF":"sigmoid","weights":[0.07030263961505323,-0.26545912729664556,0.37560221411197214,0.2923244847426285,-0.8896329641128919,0.5311148426965802,0.6182392962661842]}],[{"activationF":"sigmoid","weights":[-0.4203698112641414,0.45388143955720217,-0.13615987609385183,-0.7968643692136412,-0.8999692624692617,-0.20406638996276305]},{"activationF":"sigmoid","weights":[-0.11406885695371116,-0.18786441034717605,-0.5055219662230839,0.04524373070462362,-0.11891872316176122,-0.9069098420319777]},{"activationF":"sigmoid","weights":[-0.49189194418627435,0.9655404131465519,-0.7461112874103033,-0.2337494582987365,0.03348554207672061,0.5612844635907202]}]]"""

  test("neural network to json") {
    assertEquals(NeuralNetwork.toJson(neuralNetwork), neuralNetworkJson)
  }

  test("neural network of json") {
    assertEquals(NeuralNetwork.ofJson(neuralNetworkJson), neuralNetwork)
  }

  val inputValues = List(1.0, 0.5625, 0.511111, 0.47619)

  val prediction: List[Double] =
    NeuralNetwork.prediction(neuralNetwork, inputValues)

  test("neural network prediction") {
    assertEquals(
      prediction,
      List(0.1955367616220067, 0.29453217823829403, 0.48811285629821577)
    )
  }

  val learningRate = 0.99

  val expectedOutput = List(0.4, 0.05, 0.2)

  test("neural network normal errors") {
    assertEquals(
      NeuralNetwork.errors(neuralNetwork, learningRate, inputValues, expectedOutput),
      List(0.03084857932439276, 0.017352325869970923, 0.015767048227069706, 0.014689784988482587)
    )
  }

  test("neural network zero errors") {
    assertEquals(
      NeuralNetwork.errors(neuralNetwork, learningRate, inputValues, prediction),
      List(0, 0, 0, 0)
    )
  }

  val fitNetwork: NeuralNetwork =
    NeuralNetwork.fit(neuralNetwork, learningRate, inputValues, expectedOutput)

  test("fit neural network prediction") {
    assertEquals(
      NeuralNetwork.prediction(fitNetwork, inputValues),
      List(0.21003023182099667, 0.2720325964336182, 0.44366449684542975)
    )
  }

}

File: implementation/Elixir/synapses/test/seed_network_test.exs
defmodule SeedNetworkTest do
  use ExUnit.Case
  alias Synapses.{NeuralNetwork}

  def layers(), do: [4, 6, 5, 3]

  def neuralNetwork(), do: NeuralNetwork.initWithSeed(1000, layers())

  def inputValues(), do: [1.0, 0.5625, 0.511111, 0.47619]

  def prediction(), do: NeuralNetwork.prediction(neuralNetwork(), inputValues())

  def learningRate(), do: 0.99

  def expectedOutput(), do: [0.2, 0.8, 0.01]

  def fitNetwork(), do: NeuralNetwork.fit(neuralNetwork(), learningRate(), inputValues(), expectedOutput())

  test "neural network to json" do
    neuralNetworkJson = "[[{\"activationF\":\"sigmoid\",\"weights\":[0.97591192029471424,0.608726848593679616,-0.10449463866742392,0.86215539315831552,0.508145836217687808]},{\"activationF\":\"sigmoid\",\"weights\":[0.42884311071260608,-0.928013620271803264,-0.841588171984059648,0.6369844981782016,0.493257044371126848]},{\"activationF\":\"sigmoid\",\"weights\":[-0.250953134676056512,0.486245119456245824,0.373046333986740224,0.48124515069088416,0.437726902073087424]},{\"activationF\":\"sigmoid\",\"weights\":[0.34062889128564544,-0.075654855768376944,0.12336723424429552,-0.590345384807812224,-0.025738667762999156]},{\"activationF\":\"sigmoid\",\"weights\":[0.097662715981162496,0.102552052937866064,0.763052863780054656,0.864186513158113408,0.852234019355143168]},{\"activationF\":\"sigmoid\",\"weights\":[0.500883936382927808,-0.815786202686251648,-0.069786319095865856,0.052542488749475872,-0.42187283962216096]}],[{\"activationF\":\"sigmoid\",\"weights\":[-0.545387987433667776,0.193848784045706816,-0.894539921764575488,0.017658813141432584,-0.99557544998996224,0.78955757947190272,-0.135504802947656656]},{\"activationF\":\"sigmoid\",\"weights\":[0.7449544487116448,0.820276742068607488,-0.100644851204899856,-0.433682731359977792,-0.458571748080177216,-0.18585897382787264,0.7339677148976192]},{\"activationF\":\"sigmoid\",\"weights\":[-0.695678943892840576,0.84676260345272,0.755033002416591488,-0.147621102357376128,0.94595696395035776,0.242011207645256736,0.093684348045486752]},{\"activationF\":\"sigmoid\",\"weights\":[0.16876442707872384,-0.216468406120710048,-0.25336739896256112,0.416066890622018432,0.025323671178192784,0.517898431689628928,0.160560495194321792]},{\"activationF\":\"sigmoid\",\"weights\":[-0.295402891098786496,0.234971109993689984,0.175273896680819168,-0.616672251978913152,0.759511997533433216,0.619803202444554112,0.183035240691822976]}],[{\"activationF\":\"sigmoid\",\"weights\":[0.882362387086611968,-0.371599210719792064,-0.245399966611133504,0.03076713228196848,0.656814133282412672,-0.27558689300370864]},{\"activationF\":\"sigmoid\",\"weights\":[0.63357058507698304,-0.048198159104340776,-0.49323248769687392,-0.684966179534859392,-0.357854336273865088,0.918464864900381056]},{\"activationF\":\"sigmoid\",\"weights\":[0.593111751544443904,-0.006652794363385395,0.867715902167896192,-0.934252979757118848,0.657668916181701376,0.15566771739603104]}]]"
    assert NeuralNetwork.toJson(neuralNetwork()) == neuralNetworkJson
  end

  test "neural network of/to json" do
    neuralNetworkJson = NeuralNetwork.toJson(neuralNetwork())
    recreatedNeuralNetworkJson = neuralNetworkJson
                                 |> NeuralNetwork.ofJson
                                 |> NeuralNetwork.toJson
    assert recreatedNeuralNetworkJson == neuralNetworkJson
  end

  test "neural network prediction" do
    assert prediction() == [0.7018483008852783, 0.5232699523175631, 0.746950953587391]
  end

  test "neural network normal errors" do
    assert NeuralNetwork.errors(neuralNetwork(), learningRate(), inputValues(), expectedOutput()) == [
             0.07624623311148832,
             0.042888506125212174,
             0.0389702884518459,
             0.036307693745359616
           ]
  end

  test "neural network zero errors" do
    assert NeuralNetwork.errors(neuralNetwork(), learningRate(), inputValues(), prediction()) == [0, 0, 0, 0]
  end

  test "fit neural network prediction" do
    assert NeuralNetwork.prediction(fitNetwork(), inputValues()) == [
             0.6335205999385805,
             0.5756314596704061,
             0.6599122411687741
           ]
  end

end
